applications/IgaApplication/custom_elements/curve_base_discrete_element.cpp

```diff
@@ -0,0 +1,168 @@
+                rBaseVector[2] += rDN_De(0, i) * GetGeometry()[i].Z();
+            }
+        }
+    }
+
+    //************************************************************************************
+    //************************************************************************************
+    void CurveBaseDiscreteElement::GetBoundaryEdgeBaseVector(
+        const Matrix& DN_De,
+        const array_1d<double, 2>& Tangents,
+        Vector& rBaseVector)
+    {
+        Matrix J;
+        Jacobian(DN_De, J);
+
+        //basis vectors g1 and g2
+        array_1d<double, 3> g1;
+        array_1d<double, 3> g2;
+
+        g1[0] = J(0, 0);
+        g2[0] = J(0, 1);
+        g1[1] = J(1, 0);
+        g2[1] = J(1, 1);
+        g1[2] = J(2, 0);
+        g2[2] = J(2, 1);
+
+        rBaseVector = g1 * Tangents[0] + g2 * Tangents[1];
+    }
+
+    //************************************************************************************
+    //************************************************************************************
+    void CurveBaseDiscreteElement::Get1stVariationsAxialStrain(
+        Vector& rEpsilon1stVariationDoF,
+        const Vector& rBaseVector,
+        const int& rNumberOfDoFs,
+        const Matrix& rDN_De)
+    {
+        int mat_size = rDN_De.size1()*rNumberOfDoFs;
+
+        if (rEpsilon1stVariationDoF.size() != mat_size)
+            rEpsilon1stVariationDoF.resize(mat_size, false);
+        rEpsilon1stVariationDoF = ZeroVector(mat_size);
+
+        int xyz_r = 0;
+        int i = 0;
+        for (int r = 0; r < mat_size; r++)
+        {
+            xyz_r = r % rNumberOfDoFs; //0 ->disp_x; 1 ->disp_y; 2 ->disp_z
+            i = r / rNumberOfDoFs;     // index for the shape functions
+            if (xyz_r>2)
+                rEpsilon1stVariationDoF[r] = 0.0;
+            else
+                rEpsilon1stVariationDoF[r] = rBaseVector[xyz_r] * rDN_De(i, 0);
+        }
+    }
+
+    //************************************************************************************
+    //************************************************************************************
+    void CurveBaseDiscreteElement::Get2ndVariationsAxialStrain(
+        Matrix& rEpsilon2ndVariationDoF,
+        const int& rNumberOfDoFs,
+        const Matrix& rDN_De)
+    {
+        int mat_size = rDN_De.size1()*rNumberOfDoFs;
+
+        if ((rEpsilon2ndVariationDoF.size1() != mat_size) && (rEpsilon2ndVariationDoF.size1() != mat_size))
+            rEpsilon2ndVariationDoF.resize(mat_size, mat_size, false);
+        rEpsilon2ndVariationDoF = ZeroMatrix(mat_size, mat_size);
+
+        for (int r = 0; r<mat_size; r++) //in the case
+        {
+            int xyz_r = r % rNumberOfDoFs; //0 ->disp_x; 1 ->disp_y; 2 ->disp_z; 3 -> rot_tan
+            int i = r / rNumberOfDoFs;     // index for the shape functions
+            if (xyz_r>2)
+                for (int s = 0; s<mat_size; s++)
+                    rEpsilon2ndVariationDoF(r, s) = 0.0;
```

Comment:
@philbucher @msandre 
If we are discussing about `++i` vs. `i++` we should also start to use brackets for `if`/`for`-one-liners

Reply:
We don't need to start a C++ civil war here...

Reply:
In this case `++i` is not different of `i++`
I think it is good practice and does not hurt to use `++i` everywhere and this is what we started to use in most of Kratos in the recent months

I don't see this as a blocker, just telling you what is usually used

Reply:
was not meant seriously 😉

Reply:
The aim should always be to write better code. This means correcting bad habits (even small ones) when we see them as part of review. The common habit of writing  `i++` instead of `++i` is not best practice and is easy to correct. Time should not be wasted debating it.

The explanation can be found from the head of the c++ standards committee:
https://herbsutter.com/2013/05/13/gotw-2-solution-temporary-objects/

I don't see the comparison with the brackets in `if` statements.

Reply:
@msandre 

> was not meant seriously 😉

(btw: I am not the author of this file)

The discussion of the article is about using `i++` with iterators. I completely agree with that and I always use `++it` for iterators. I usually use `i++` in for-loops because most of my editors have a automatic feature for inserting loops and most of the languages I am using don't have iterators but loops.

I also agree with the aim of writing better code. I just want to say that leaving the brackets away is also a very bad habit:

```cpp
if (true)
    if (false)
        std::cout << "A";
else
    std::cout << "B";

>> B
```

I am coding now for over 19 years in different languages. I NEVER had problems with `i++` but I actually HAD PROBLEMS with the brackets in various projects. So I just want to say that we **maybe** should also take a look on this.

Take a look at the examples in [your reference](https://herbsutter.com/2013/05/13/gotw-2-solution-temporary-objects/): Mr. Sutter also uses brackets for one-liner.

Reply:
Nice example :+1: